### PR TITLE
【BIT】转换能力补齐：paddle.nn.functional.conv[1 2 3]d_transpose

### PR DESCRIPTION
**Pull request by cangtianhuang**
### 1. [paddle.nn.functional.conv1d_transpose](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/functional/conv1d_transpose_cn.html)

查询 [torch.nn.functional.conv_transpose1d 与 paddle.nn.functional.conv1d_transpose 对照表](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/functional/torch.nn.functional.conv_transpose1d.html) ，发现是 `paddle 参数更多`，注册并编写 Conv1dTransposeRule 类

查阅文档发现，paddle 参数更多：paddle 多支持 data_format 参数，使用 permute 调换输入与输出维度顺序；paddle 多支持 output_size 参数，阅读源码与广泛测试后发现，output_size 影响的是 output_padding 而非 padding，因此计算默认 L_out，则 output_padding = output_size - L_out

paddle 与 torch 的多个参数用法不同：stride、output_padding、dilation、output_size 需转为 int 类型；bias 参数需手动广播至 out_channels 大小，out_channels 可通过 weight.size(1) * groups 得到

此外，padding 参数具有非常丰富的形式：当 padding 为 “SAME” 时，计算所需的 padding 大小；padding 为 “VALID” 时，padding 设置为 0；padding 为长度为 1 的列表时，直接提取；padding 为长度为 2 的列表时，代表非对称填充，torch 对应的 api 不支持非对称填充，因此需记录 crop 大小，对 torch 的输出进行手动裁剪（此为 transpose api，因此是对输出裁剪而非对输入填充）；padding 为长度为 3、元素为长度为 2 的嵌套列表时，代表对输入的所有维度的非对称填充，因此根据 data_format 提取并记录 crop 大小，对 torch 输出进行手动裁剪

### 2. [paddle.nn.functional.conv2d_transpose](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/functional/conv2d_transpose_cn.html)

查询 [torch.nn.functional.conv_transpose2d 与 paddle.nn.functional.conv2d_transpose 对照表](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/functional/torch.nn.functional.conv_transpose2d.html) ，发现是 `paddle 参数更多`，注册并编写 Conv2dTransposeRule 类

与 paddle.nn.functional.conv1d_transpose 类似，但是参数类型略有不同，维度更多，区别如下：

paddle 多支持的 output_size 参数需分别对 H、W 维度计算默认 L_out，设置 output_padding[i] = output_size[i] - L_out，转化为 tuple 类型

stride、output_padding、dilation 若为列表，需转为 tuple 类型

当 padding 为 “SAME” 时，分别计算 H、W 维度所需的 padding 大小；padding 为 “VALID” 时，padding 设置为 0；padding 为长度为 2 的列表时，转化为 tuple 类型；padding 为长度为 4 的列表时，代表非对称填充，记录 crop 大小，对 torch 的输出进行手动裁剪；padding 为长度为 4、元素为长度为 2 的嵌套列表时，代表对输入的所有维度的非对称填充，因此根据 data_format 提取并记录 crop 大小，对 torch 输出进行手动裁剪

### 3. [paddle.nn.functional.conv3d_transpose](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/functional/conv3d_transpose_cn.html)

查询 [torch.nn.functional.conv_transpose3d 与 paddle.nn.functional.conv3d_transpose 对照表](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/functional/torch.nn.functional.conv_transpose3d.html) ，发现是 `paddle 参数更多`，注册并编写 Conv3dTransposeRule 类

与 paddle.nn.functional.conv2d_transpose 类似，但是维度更多，区别如下：

当 padding 为 “SAME” 时，分别计算 D、H、W 维度所需的 padding 大小；padding 为 “VALID” 时，padding 设置为 0；padding 为长度为 3 的列表时，转化为 tuple 类型；padding 为长度为 6 的列表时，代表非对称填充，记录 crop 大小，对 torch 的输出进行手动裁剪；padding 为长度为 5、元素为长度为 2 的嵌套列表时，代表对输入的所有维度的非对称填充，因此根据 data_format 提取并记录 crop 大小，对 torch 输出进行手动裁剪

### paddle.nn.functional.avg_pool3d

偶然发现 AvgPoolRule 的 permute 顺序错误，回测后发现了两个报错，因此修复了 permute 顺序、增加手动 pad，测试全部通过

## 测试结果

paddle.nn.functional.conv[1 2 3]d_transpose 配置均通过 accuracy 测试，移入 support2torch_1.txt 中

image

### CODE REVIEW

**Review by Cutelemon6 — APPROVED** [review]
LGTM

**Review by wanghuancoder — APPROVED** [review]
LGTM
